### DISCUSSION TITLE
Path() helper mangles escape codes

### DISCUSSION THREAD

**Discussion by tmm1** [earthboundkid/requests#95] [Bugs] [answered]
if i use `Path("/groups/test%2Ffoo")` which is the actual URL used by a legacy service, the request that's sent ends up being `/groups/test%252Ffoo` which the remote server doesn't recognize

this is likely because of this:

https://github.com/carlmjohnson/requests/blob/5d11526188c82fefcbdbde64537a4ad49841cdff/core_url.go#L75-L78

**Comment by tmm1** [comment, marked as answer]
As a workaround I appended the path to the url and passed to `BaseURL()` instead.

**Reply by earthboundkid** [reply]
Isn’t %2F space? What happens if you do `Path("/groups/test foo")`?

**Reply by tmm1** [reply]
`%2F` is an escaped `/`

**Comment by Alan3344** [comment]
❓ can only be passed in the URL for splicing. If they are escaped in the `Path` and `Pathf` methods, I am very distressed.

```go
package main

import (
	"context"
	"log"

	"github.com/carlmjohnson/requests"
)

func main() {
	var err error
	var rawBody string
	if err = requests.URL("https://httpbun.org").
		Path("/get?n=aaa&n=www").
		// Pathf("/get?n=%s&n=%s", "aaa", "www").
		// Pathf("/get%s", "?n=aaa&n=www").
		// Param("n", "aaa", "www").
		ToString(&rawBody).
		Fetch(context.Background()); err != nil {
		log.Println("[fetch]", err)
	}
	log.Println(rawBody)
}
```

Output:
```bash
[fetch] ErrValidator: response error for https://httpbun.org/get%3Fn=aaa&n=www: unexpected status: 404
```

In order to unify the writing, I did not use `Param` to pass `Args`.

Hopefully it would be smarter and not escape the first ❓ after `Path`.

Thanks for your hard work!

**Reply by earthboundkid** [reply]
[Here](https://go.dev/play/p/ArH-xICwZdf) is a minimal example:

```go
u, _ := requests.URL("https://example.com").
	Path("/get?n=aaa&n=www").
	URL()
fmt.Println(u.String()) // Outputs https://example.com/get%3Fn=aaa&n=www
```

This is expected behavior because `Path` sets the path and not the query. It would be unsafe if, for example, you tried to request `Path("/user/"+userName)` and the user could trick you into sending a query string instead.

To get a query, [do this](https://go.dev/play/p/BO9LqSZirG6):

```go
s := "/get?n=aaa&n=www"
u, _ := requests.URL("https://example.com" + s).
	URL()
fmt.Println(u.String()) // Outputs https://example.com/get?n=aaa&n=www
```

**Reply by Alan3344** [reply]
I think both rules should be allowed for Path matching.
```
(\w+/)+\?|(\w+/)+(\w+)\?
```
![image](https://github.com/user-attachments/assets/516f498f-e180-4898-9c9a-4bd34df37f37)

Html demo
```html
<p>path/to/u/api/get/?n=aaa</p>
<p>path/to/u/api/get?n=aaa</p>
```
<p>path/to/u/api/get/?n=aaa</p>
<p>path/to/u/api/get?n=aaa</p>

What do you think?

**Reply by earthboundkid** [reply]
I think what I said before, "This is expected behavior because Path sets the path and not the query."
